### PR TITLE
BUG: Unstack/pivot raising ValueError on large result

### LINKED ISSUE #26314: Pivot / unstack on large data frame does not work int32 overflow
#### Code Sample, a copy-pastable example if possible

```python
predictor_purchases_p = predictor_purchases.groupby(["ag", "artikelnr"])["som"].max().unstack().fillna(0)

or

predictor_purchases_p = predictor_purchases.pivot(index="ag", columns="artikelnr", value="som")
```
#### Problem description
I'm working on rather large data (>100GB in memory) on a beefy server (3TB ram)
When refactoring my code from pandas 0.21 to latest version, the pivot / unstack now returns an exception.

Unstacked DataFrame is too big, causing int32 overflow

I was able to eliminate the problem by changing the reshape.py:
modify line 121from dtype np.int32 to dtype np.int64:
 `num_cells = np.multiply(num_rows, num_columns, dtype=np.int64)`

#### Expected Output

Not being limited by int32 dims when reshaping a data frame.
This feels like a restriction which should not be there.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.3.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-862.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.2
pytest: None
pip: 19.1
setuptools: 41.0.1
Cython: 0.29.7
numpy: 1.16.3
scipy: 1.2.1
pyarrow: None
xarray: 0.12.1
IPython: 7.5.0
sphinx: None
patsy: 0.5.1
dateutil: 2.8.0
pytz: 2019.1
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 3.0.3
openpyxl: 2.6.2
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: None
bs4: None
html5lib: None
sqlalchemy: 1.3.3
pymysql: None
psycopg2: None
jinja2: 2.10.1
s3fs: None
fastparquet: 0.3.0
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #26314
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/reshape/test_pivot.py [hunk 1]
```diff
@@ -1991,12 +1993,13 @@ def test_pivot_string_func_vs_func(self, f, f_numpy):
...
         # GH 20601
+        # GH 26314: Change ValueError to PerformanceWarning
         df = DataFrame(
             {"ind1": np.arange(2 ** 16), "ind2": np.arange(2 ** 16), "count": 0}
         )
 
-        msg = "Unstacked DataFrame is too big, causing int32 overflow"
-        with pytest.raises(ValueError, match=msg):
+        msg = "The following operation may generate"
+        with tm.assert_produces_warning(PerformanceWarning, match=msg):
```
Comment (jreback) [inline pandas/tests/reshape/test_pivot.py:2002]: does this test actually run?

Comment (mroeschke) [reply pandas/tests/reshape/test_pivot.py:2002]: Yup. These are running on the slow builds.

```
6.56s call     pandas/tests/frame/test_stack_unstack.py::TestStackUnstackMultiLevel::test_unstack_number_of_levels_larger_than_int32
6.50s call     pandas/tests/reshape/test_pivot.py::TestPivotTable::test_pivot_number_of_levels_larger_than_int32

```

**Comment by jreback** [comment]
thanks @mroeschke

**Comment by simonjayhawkins** [comment]
@mroeschke this seems to significantly increase the run time running the tests locally.

also the tests fail with pytest-xdist... node down.

**Comment by mroeschke** [comment]
> @mroeschke this seems to significantly increase the run time running the tests locally.
> 
> also the tests fail with pytest-xdist... node down.

Addressed in https://github.com/pandas-dev/pandas/pull/45112. Now takes ~4 seconds locally.

**Pull request merged**